File: BookLand_FE/src/types/PaymentTransaction.ts
import type { Bill } from './Bill';
import type { PaymentMethod } from './PaymentMethod';

export const TransactionStatus = {
    PENDING: 'PENDING',
    SUCCESS: 'SUCCESS',
    FAILED: 'FAILED'
} as const;

export type TransactionStatus = (typeof TransactionStatus)[keyof typeof TransactionStatus];

export interface PaymentTransaction {
    id: number;
    bill: Bill;
    paymentMethod: PaymentMethod;
    provider?: string;
    amount: number;
    transactionCode?: string;
    providerTransactionId?: string;
    status: TransactionStatus;
    payUrl?: string;
    responseCode?: string;
    responseMessage?: string;
    paidAt?: string;
    createdAt?: string;
}

File: services/order-service/src/main/java/com/bookland/order/entity/PaymentTransaction.java
package com.bookland.order.entity;

import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "payment_transaction")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class PaymentTransaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "bill_id", nullable = false)
    private Bill bill;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "payment_method_id", nullable = false)
    private PaymentMethod paymentMethod;

    private String provider;

    @Column(nullable = false)
    private Double amount;

    @Column(name = "transaction_code", unique = true)
    private String transactionCode;

    @Column(name = "provider_transaction_id")
    private String providerTransactionId;

    @Enumerated(EnumType.STRING)
    @Builder.Default
    private TransactionStatus status = TransactionStatus.PENDING;

    @Column(name = "pay_url", columnDefinition = "TEXT")
    private String payUrl;

    @Column(name = "response_code")
    private String responseCode;

    @Column(name = "response_message", columnDefinition = "TEXT")
    private String responseMessage;

    @Column(name = "paid_at")
    private LocalDateTime paidAt;

    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }

    public enum TransactionStatus {
        PENDING, SUCCESS, FAILED
    }
}
